docs/source/concepts/InstrumentAccessLayers.rst

```diff
@@ -31,216 +33,170 @@ ____________
-
-Useful Tips
-___________
-
-* Creation of ``SpectrumInfo`` objects is not cheap. Make sure ``ws.spectrumInfo()`` is called outside of loops that go over all spectra.
-* If a ``SpectrumInfo`` object is required for more than one workspace then include the workspace name in the name of the ``SpectrumInfo`` object, to avoid confusion.
-* Get the ``SpectrumInfo`` object as a const reference and use auto - ``const auto &spectrumInfo = ws->spectrumInfo();``.
-* Do not forget to add the import - ``#include "MantidAPI/SpectrumInfo.h"``.
-
-Complete Examples
-_________________
-
-* `FindCenterOfMassPosition2.cpp <https://github.com/mantidproject/mantid/pull/17394/files#diff-905c244467474fc320eaf3b8c7a9f0dd>`_
-
-* `CreatePSDBleedMask.cpp <https://github.com/mantidproject/mantid/pull/18218/files#diff-f490acf06e76f93898dc7d486c8dfa93>`_
-
-* `HRPDSlabCanAbsorption.cpp <https://github.com/mantidproject/mantid/pull/18464/files#diff-fc151838d9d7cc2e4ea469e98472c791>`_
-
-DetectorInfo
-############
+**After - looping over index**
 
-Basics
-______
+.. code-block:: c++
 
-The conversion is similar to that for ``SpectrumInfo``. For ``DetectorInfo`` all instances of ``Instrument::getDetector()`` should be replaced using calls to the ``DetectorInfo`` object obtained from ``MatrixWorkspace::detectorInfo()``. The methods below can then be called on ``DetectorInfo`` instead of on the detector.
+  #include "MantidGeometry/Instrument/ComponentInfo.h"
 
-* ``isMonitor(detIndex)``
-* ``isMasked(detIndex)``
-* ``l2(detIndex)``
-* ``twoTheta(detIndex)``
-* ``signedTwoTheta(detIndex)``
-* ``position(detIndex)``
+  ...
 
-**Indexing**
+  const auto &componentInfo = ws->componentInfo();
+  if (componentInfo.size() == 0)
+    throw runtime_error("There is no instrument in input workspace.")
 
-The ``DetectorInfo`` object is accessed by an index going from 0 to the number of detectors. A vector of detector IDs can be obtained from a call to ``detectorInfo.detectorIDs()``.
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    componentInfo.solidAngle(i, observer);
+  }
 
-It is also possible to use the method ``detectorInfo.indexOf(detID)`` to get the index for a particular detector ID. However, this is a call to a lookup in an unordered map, so is an expensive calculation which should be avoided where possible.
+Access to Components and working with Detectors
+_______________________________________________
 
-Below is an example refactoring.
+Detector Indices are the same as the corresponding Component Indices. Note that there are no dynamic casts.
 
-**Before refactoring**
+**Combining DetectorInfo and ComponentInfo**
 
 .. code-block:: c++
 
-  auto instrument = ws->getInstrument();
-  if (!instrument)
-    throw runtime_error("There is no instrument in input workspace.")
```

Comment:
A workspace always has an instrument, so the check in the old code is pointless.